executor/cook/io_helper.py

```diff
@@ -0,0 +1,143 @@
+#!/usr/bin/env python3
+
+"""This module ensures atomic writes to stdout and stderr.
+"""
+
+import logging
+import os
+import sys
+from threading import Lock, Thread
+
+__stdout_lock__ = Lock()
+__stderr_lock__ = Lock()
+
+
+def print_out(string_data, flush=False, newline=True):
+    """Wrapper function that prints to stdout in a thread-safe manner using the __stdout_lock__ lock.
+
+    Parameters
+    ----------
+    string_data: string
+        The string to output
+    flush: boolean
+        Flag determining whether to trigger a sys.stdout.flush()
+    newline: boolean
+        Flag determining whether to output a newline at the end
+
+    Returns
+    -------
+    Nothing.
+    """
+    message = '{}{}'.format(string_data, os.linesep) if newline else string_data
+    with __stdout_lock__:
+        sys.stdout.write(message)
+    if flush:
+        sys.stdout.flush()
```

Comment:
Why shouldn't flush be protected by the the lock?

Reply:
I want to reduce the work done in the critical section. Do you see any harm in having it outside and potentially running flush in parallel?

Reply:
I couldn't get two `sys.stdout.write` calls to interleave.  If the newlines were inside the buffer being written, we might be able to avoid these locks entirely.

I can't find any docs indicating whether `flush()` is thread-safe or not.  Since we're perhaps being overprotective with the write locks, we could also be overprotective on the `flush()` calls.  It still feels odd to me that one is protected and one isn't.

Reply:
Maybe `sys.stdout` calls aren't thread-safe after all:

<pre>
>>> import sys
>>> sys.stdout
<_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>
>>>
</pre>

From https://docs.python.org/3/library/io.html:

> 16.2.4.3. Multi-threading
FileIO objects are thread-safe to the extent that the operating system calls (such as read(2) under Unix) they wrap are thread-safe too.

> Binary buffered objects (instances of BufferedReader, BufferedWriter, BufferedRandom and BufferedRWPair) protect their internal structures using a lock; it is therefore safe to call them from multiple threads at once.

> TextIOWrapper objects are not thread-safe.